src/createApp.ts

```diff
@@ -569,17 +569,28 @@ function addIdentifier(app: App, id: Identifier) {
 
 function createCustomWidget(app: App, id: string) {
 	const { registryProvider, defaultWidgetStore } = app;
+	let factoryHandle: Handle;
 	// istanbul ignore if
 	if (!defaultWidgetStore) {
 		throw new Error('A default widget store must be configured in order to create custom widgets');
 	}
 
 	return defaultWidgetStore.get(id).then((state: any) => {
-		const options: WidgetFactoryOptions = { id, stateFrom: defaultWidgetStore, registryProvider, state };
 		const customFactory = customElementFactories.get(app).get(state.type);
-		return customFactory(options);
+		const factories = widgetFactories.get(app);
+		let factory: RegisteredFactory<WidgetLike>;
+		try {
+			// add final check before attempting to register the factory
+			factory = factories.get(id);
+		}
+		catch (err) {
+			factoryHandle = app.registerWidgetFactory(id, customFactory);
+			factory = factories.get(id);
+		}
+
+		return factory();
```

Comment:
And return `app.getWidget(id)` here. Works whether you need to use an existing factory or the freshly registered custom factory.